File: PalmBee/pb_control/scripts/control_manager.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from mavros_msgs.srv import WaypointPush, WaypointClear
from mavros_msgs.msg import Waypoint
from sensor_msgs.msg import NavSatFix


class EKFOriginMissionUploader(Node):

    def __init__(self):
        super().__init__('ekf_origin_mission_uploader')

        self.origin_received = False

        # Subscribe to EKF origin
        self.origin_sub = self.create_subscription(
            NavSatFix,
            '/mavros/global_position/gp_origin',
            self.origin_cb,
            10
        )

        # Mission services
        self.clear_cli = self.create_client(
            WaypointClear,
            '/mavros/mission/clear'
        )
        self.push_cli = self.create_client(
            WaypointPush,
            '/mavros/mission/push'
        )

        self.get_logger().info('Waiting for EKF origin...')

    def origin_cb(self, msg: NavSatFix):
        if self.origin_received:
            return

        self.origin_received = True
        self.get_logger().info(
            f"EKF origin received (lat={msg.latitude}, lon={msg.longitude})"
        )

        self.upload_mission()

    def upload_mission(self):
        while not self.clear_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for mission clear service...')

        while not self.push_cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for mission push service...')

        takeoff_alt = 1.5    # meters
        move_dist = 2.0      # meters
        loiter_time = 2.0    # seconds

        waypoints = []

        # 0️⃣ TAKEOFF
        wp = Waypoint()
        wp.frame = Waypoint.FRAME_LOCAL_NED
        wp.command = Waypoint.NAV_TAKEOFF
        wp.is_current = True
        wp.autocontinue = True
        wp.param7 = takeoff_alt
        waypoints.append(wp)

        # 1️⃣ MOVE 2m FROM ORIGIN
        wp = Waypoint()
        wp.frame = Waypoint.FRAME_LOCAL_NED
        wp.command = Waypoint.NAV_WAYPOINT
        wp.is_current = False
        wp.autocontinue = True
        wp.x_lat = move_dist     # North
        wp.y_long = 0.0          # East
        wp.z_alt = -takeoff_alt # Up
        waypoints.append(wp)

        # 2️⃣ HOVER / LOITER
        wp = Waypoint()
        wp.frame = Waypoint.FRAME_LOCAL_NED
        wp.command = Waypoint.NAV_LOITER_TIME
        wp.is_current = False
        wp.autocontinue = True
        wp.param1 = loiter_time
        wp.x_lat = move_dist
        wp.y_long = 0.0
        wp.z_alt = -takeoff_alt
        waypoints.append(wp)

        # 3️⃣ LAND
        wp = Waypoint()
        wp.frame = Waypoint.FRAME_LOCAL_NED
        wp.command = Waypoint.NAV_LAND
        wp.is_current = False
        wp.autocontinue = True
        waypoints.append(wp)

        # Clear mission
        self.clear_cli.call_async(WaypointClear.Request())

        # Upload mission
        req = WaypointPush.Request()
        req.waypoints = waypoints

        future = self.push_cli.call_async(req)
        future.add_done_callback(self.upload_result)

    def upload_result(self, future):
        result = future.result()
        if result.success:
            self.get_logger().info(
                f"Mission uploaded successfully ({result.wp_transfered} waypoints)"
            )
        else:
            self.get_logger().error("Mission upload FAILED")


def main():
    rclpy.init()
    node = EKFOriginMissionUploader()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: PalmBee/pb_control/src/control_manager.cpp
#include <memory>
#include <vector>
#include <cmath>

#include "rclcpp/rclcpp.hpp"

#include "mavros_msgs/msg/command_code.hpp"
#include "mavros_msgs/srv/waypoint_clear.hpp"
#include "mavros_msgs/srv/waypoint_push.hpp"
#include "mavros_msgs/msg/waypoint.hpp"

#include "geographic_msgs/msg/geo_point_stamped.hpp"

using std::placeholders::_1;

// MAVLink numeric frame (portable, Humble-safe)
static constexpr uint8_t MAV_FRAME_GLOBAL_RELATIVE_ALT = 3;

class EKFOriginMissionUploader : public rclcpp::Node
{
public:
  EKFOriginMissionUploader()
  : Node("ekf_origin_mission_uploader"),
    origin_received_(false)
  {
    auto qos = rclcpp::QoS(rclcpp::KeepLast(1))
                 .reliable()
                 .transient_local();

    origin_sub_ = this->create_subscription<
      geographic_msgs::msg::GeoPointStamped>(
        "/mavros/global_position/gp_origin",
        qos,
        std::bind(&EKFOriginMissionUploader::originCallback, this, _1)
      );

    clear_client_ = this->create_client<mavros_msgs::srv::WaypointClear>(
      "/mavros/mission/clear"
    );

    push_client_ = this->create_client<mavros_msgs::srv::WaypointPush>(
      "/mavros/mission/push"
    );

    RCLCPP_INFO(this->get_logger(), "Waiting for EKF origin...");
  }

private:
  // ===== EKF ORIGIN =====
  bool   origin_received_;
  double origin_lat_;
  double origin_lon_;
  double origin_alt_;

  void originCallback(
    const geographic_msgs::msg::GeoPointStamped::SharedPtr msg)
  {
    if (origin_received_) return;

    origin_received_ = true;

    origin_lat_ = msg->position.latitude;
    origin_lon_ = msg->position.longitude;
    origin_alt_ = msg->position.altitude;

    RCLCPP_INFO(
      this->get_logger(),
      "EKF origin received (lat=%.8f, lon=%.8f, alt=%.2f)",
      origin_lat_, origin_lon_, origin_alt_
    );

    uploadMission();
  }

  // ===== METERS → LAT/LON =====
  void metersToLatLon(
    double north_m,
    double east_m,
    double &lat_out,
    double &lon_out)
  {
    constexpr double R = 6378137.0; // Earth radius (WGS84)
    double lat_rad = origin_lat_ * M_PI / 180.0;

    lat_out = origin_lat_ + (north_m / R) * 180.0 / M_PI;
    lon_out = origin_lon_ + (east_m  / (R * cos(lat_rad))) * 180.0 / M_PI;
  }

  // ===== MISSION =====
  void uploadMission()
  {
    const float takeoff_alt = 1.5f;
    const float move_dist   = 2.0f;
    const float loiter_time = 2.0f;

    std::vector<mavros_msgs::msg::Waypoint> waypoints;

    double lat, lon;

    // WP0 — climb at EKF origin
    {
      mavros_msgs::msg::Waypoint wp;
      wp.frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
      wp.command = mavros_msgs::msg::CommandCode::NAV_WAYPOINT;
      wp.is_current = true;
      wp.autocontinue = true;
      wp.x_lat  = origin_lat_;
      wp.y_long = origin_lon_;
      wp.z_alt  = takeoff_alt;
      waypoints.push_back(wp);
    }

    // WP1 — move FORWARD (EAST, because yaw ≈ 90°)
    metersToLatLon(0.0, move_dist, lat, lon);

    {
      mavros_msgs::msg::Waypoint wp;
      wp.frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
      wp.command = mavros_msgs::msg::CommandCode::NAV_WAYPOINT;
      wp.autocontinue = true;
      wp.x_lat  = lat;
      wp.y_long = lon;
      wp.z_alt  = takeoff_alt;
      waypoints.push_back(wp);
    }

    // WP2 — loiter
    {
      mavros_msgs::msg::Waypoint wp;
      wp.frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
      wp.command = mavros_msgs::msg::CommandCode::NAV_LOITER_TIME;
      wp.autocontinue = true;
      wp.param1 = loiter_time;
      wp.x_lat  = lat;
      wp.y_long = lon;
      wp.z_alt  = takeoff_alt;
      waypoints.push_back(wp);
    }

    // WP3 — land
    {
      mavros_msgs::msg::Waypoint wp;
      wp.frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
      wp.command = mavros_msgs::msg::CommandCode::NAV_LAND;
      wp.autocontinue = true;
      wp.x_lat  = lat;
      wp.y_long = lon;
      wp.z_alt  = 0.0;
      waypoints.push_back(wp);
    }

    // CLEAR → PUSH (async, no wait)
    auto clear_req =
      std::make_shared<mavros_msgs::srv::WaypointClear::Request>();

    clear_client_->async_send_request(
      clear_req,
      [this, waypoints](
        rclcpp::Client<mavros_msgs::srv::WaypointClear>::SharedFuture)
      {
        RCLCPP_INFO(
          this->get_logger(),
          "Mission cleared, pushing GLOBAL_RELATIVE_ALT mission"
        );
        pushMission(waypoints);
      }
    );
  }

  void pushMission(
    const std::vector<mavros_msgs::msg::Waypoint>& waypoints)
  {
    auto push_req =
      std::make_shared<mavros_msgs::srv::WaypointPush::Request>();
    push_req->waypoints = waypoints;

    push_client_->async_send_request(
      push_req,
      std::bind(
        &EKFOriginMissionUploader::uploadResult,
        this,
        std::placeholders::_1
      )
    );
  }

  void uploadResult(
    rclcpp::Client<mavros_msgs::srv::WaypointPush>::SharedFuture future)
  {
    auto result = future.get();

    if (result->success) {
      RCLCPP_INFO(
        this->get_logger(),
        "Mission uploaded successfully (%u waypoints)",
        result->wp_transfered
      );
    } else {
      RCLCPP_ERROR(this->get_logger(), "Mission upload FAILED");
    }
  }

  // ROS interfaces
  rclcpp::Subscription<
    geographic_msgs::msg::GeoPointStamped>::SharedPtr origin_sub_;
  rclcpp::Client<mavros_msgs::srv::WaypointClear>::SharedPtr clear_client_;
  rclcpp::Client<mavros_msgs::srv::WaypointPush>::SharedPtr push_client_;
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<EKFOriginMissionUploader>());
  rclcpp::shutdown();
  return 0;
}
